src/librustc/hir/check_attr.rs

```diff
@@ -189,17 +194,85 @@ impl<'a, 'tcx> CheckAttrVisitor<'a, 'tcx> {
                        "conflicting representation hints");
         }
     }
+
+    fn emit_repr_error(
+        &self,
+        hint_span: Span,
+        label_span: Span,
+        hint_message: &str,
+        label_message: &str,
+    ) {
+        struct_span_err!(self.tcx.sess, hint_span, E0517, "{}", hint_message)
+            .span_label(label_span, label_message)
+            .emit();
+    }
+
+    fn check_stmt_attributes(&self, stmt: &hir::Stmt) {
+        // When checking statements ignore expressions, they will be checked later
+        if let hir::Stmt_::StmtDecl(_, _) = stmt.node {
+            for attr in stmt.node.attrs() {
+                if attr.check_name("inline") {
+                    self.check_inline(attr, &stmt.span, Target::Statement);
+                }
+                if attr.check_name("repr") {
+                    self.emit_repr_error(
+                        attr.span,
+                        stmt.span,
+                        &format!("attribute should not be applied to statements"),
+                        &format!("not a struct, enum or union"),
+                    );
+                }
+            }
+        }
+    }
+
+    fn check_expr_attributes(&self, expr: &hir::Expr) {
+        use hir::Expr_::*;
+        match expr.node {
+            // Assignments, Calls and Structs were handled by Items and Statements
+            ExprCall(..) |
+            ExprAssign(..) |
+            ExprMethodCall(..) |
+            ExprStruct(..) => return,
```

Comment:
I don't understand what is the logic here.
How attributes on `#[attr] my_fn_call()` could be handles by items or statements?

Reply:
I think I got confused reading the comments on the `Expr_` enum. I think only `ExprAssign`  overlaps with statements then.

`let a = 1` seems to be both an expression and statement?

I can take this section out and everything seems to work fine still. The error that appears for 
```rust 
#[repr]
let _y = "123";
```
 is `attribute should not be applied to statements` with and without this match.

Reply:
`let a = 1` is not an expression, `ExprAssign` is `a = b`.

If you have a statement with one expression, like
```
a = b;
```
then it's represented roughly like `StmtExpr(ExprAssign)` and attributes are attached to the outer statement and not inner expression.

Reply:
Oh Ok. I see. Thanks for the clarification.